File: app/src/main/java/com/kota/Bahamut/Command/BahamutCommandLoadLastBlock.java
package com.kota.Bahamut.Command;

import com.kota.Bahamut.ListPage.TelnetListPage;
import com.kota.Bahamut.ListPage.TelnetListPageBlock;
import com.kota.Telnet.Reference.TelnetKeyboard;
import com.kota.Telnet.TelnetOutputBuilder;

public class BahamutCommandLoadLastBlock extends TelnetCommand {

    public enum OperationMode {
        End,
        Left_Right_End,
        Home_End,
        Left_S_End,
        NotAvailable
    }

    public BahamutCommandLoadLastBlock() {
        this.Action = LoadLastBlock;
    }

    private OperationMode getLoadLastBlockMode(TelnetListPage aListPage) {

        switch (aListPage.getListType()) {
            case 1: // BoardLinkPage
                return OperationMode.Left_S_End;
            case 2: // BoardSearchPage
                if (aListPage.getSelectedIndex() != aListPage.getItemSize()) {
                    return OperationMode.End;
                }
                if (aListPage.getSelectedIndex() > 1) {
                    return OperationMode.Home_End;
                }
                return OperationMode.NotAvailable;
            default: // TelnetListPage
                // 目前的文章index != 所有文章
                if (aListPage.getSelectedIndex() != aListPage.getItemSize()) {
                    return OperationMode.End;
                }
                // 只有一篇 看板/文章
                if (aListPage.getItemSize() == 1) {
                    return OperationMode.Left_Right_End;
                }
                return OperationMode.Home_End;
        }
    }

    public void execute(TelnetListPage aListPage) {
        if (aListPage == null) {
            setDone(true);
            return;
        }
        switch (getLoadLastBlockMode(aListPage)) {
            case Left_Right_End:
                TelnetOutputBuilder.create()
                        .pushKey(TelnetKeyboard.LEFT_ARROW)
                        .pushKey(TelnetKeyboard.RIGHT_ARROW)
                        .pushKey(TelnetKeyboard.END).sendToServer();
                return;
            case Home_End:
                TelnetOutputBuilder.create()
                        .pushKey(TelnetKeyboard.HOME)
                        .pushKey(TelnetKeyboard.END).sendToServer();
                return;
            case Left_S_End:
                TelnetOutputBuilder.create()
                        .pushKey(TelnetKeyboard.LEFT_ARROW)
                        .pushKey(TelnetKeyboard.BACK_ONE_CHAR)
                        .pushKey(TelnetKeyboard.END).sendToServer();
                return;
            case End:
                TelnetOutputBuilder.create()
                        .pushKey(TelnetKeyboard.END).sendToServer();
                return;
            default:
                setDone(true);
        }
    }

    public void executeFinished(TelnetListPage aListPage, TelnetListPageBlock aPageData) {
        if (aListPage.getItemSize() > aPageData.maximumItemNumber) {
            aListPage.setItemSize(0);
            aListPage.cleanAllItem();
        }
        setDone(true);
    }

    public String toString() {
        return "[LoadLastBlock]";
    }
}

File: app/src/main/java/com/kota/Bahamut/command/BahamutCommandLoadLastBlock.kt
package com.kota.Bahamut.command

import com.kota.Bahamut.listPage.TelnetListPage
import com.kota.Bahamut.listPage.TelnetListPageBlock
import com.kota.Bahamut.pages.boardPage.BoardPageAction
import com.kota.telnet.reference.TelnetKeyboard
import com.kota.telnet.TelnetOutputBuilder.Companion.create

open class BahamutCommandLoadLastBlock : TelnetCommand() {
    enum class OperationMode {
        End,
        LeftRightEnd,
        HomeEnd,
        LeftSEnd,
        NotAvailable
    }

    init {
        this.action = BahamutCommandDef.Companion.LOAD_LAST_BLOCK
    }

    private fun getLoadLastBlockMode(aListPage: TelnetListPage): OperationMode {
        when (aListPage.listType) {
            BoardPageAction.LINK_TITLE -> return OperationMode.LeftSEnd
            BoardPageAction.SEARCH, BoardPageAction.ESSENCE -> {
                if (aListPage.selectedIndex != aListPage.listCount) {
                    return OperationMode.End
                }
                if (aListPage.selectedIndex > 1) {
                    return OperationMode.HomeEnd
                }
                return OperationMode.NotAvailable
            }

            else -> {
                // 目前的文章index != 所有文章
                if (aListPage.selectedIndex != aListPage.listCount) {
                    return OperationMode.End
                }
                // 只有一篇 看板/文章
                if (aListPage.listCount == 1) {
                    return OperationMode.LeftRightEnd
                }
                return OperationMode.HomeEnd
            }
        }
    }

    override fun execute(telnetListPage: TelnetListPage) {
        when (getLoadLastBlockMode(telnetListPage)) {
            OperationMode.LeftRightEnd -> {
                create()
                    .pushKey(TelnetKeyboard.LEFT_ARROW)
                    .pushKey(TelnetKeyboard.RIGHT_ARROW)
                    .pushKey(TelnetKeyboard.END).sendToServer()
                return
            }

            OperationMode.HomeEnd -> {
                create()
                    .pushKey(TelnetKeyboard.HOME)
                    .pushKey(TelnetKeyboard.END).sendToServer()
                return
            }

            OperationMode.LeftSEnd -> {
                create()
                    .pushKey(TelnetKeyboard.LEFT_ARROW)
                    .pushKey(TelnetKeyboard.BACK_ONE_CHAR)
                    .pushKey(TelnetKeyboard.END).sendToServer()
                return
            }

            OperationMode.End -> {
                create()
                    .pushKey(TelnetKeyboard.END).sendToServer()
                return
            }

            else -> isDone = true
        }
    }

    override fun executeFinished(telnetListPage: TelnetListPage, telnetListPageBlock: TelnetListPageBlock?) {
        if (telnetListPage.listCount > telnetListPageBlock!!.maximumItemNumber) {
            telnetListPage.listCount = 0
            telnetListPage.cleanAllItem()
        }
        isDone = true
    }

    override fun toString(): String {
        return "[LoadLastBlock]"
    }
}
